### PR TITLE
BUG: NDFrame.__setattr__ calls object.__setattribute__ on the same name, which can cause unexpected behavior

### LINKED ISSUE #56793: BUG: NDFrame.__setattr__ calls object.__setattribute__ on the same name, which can cause unexpected behavior (rewrote #56773)
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```
python
from pandas import DataFrame
from functools import cached_property
class Frame(DataFrame):
    @cached_property
    def spam(self):
        print('<time wasting task>')

Frame().spam = 5
```

```
<time wasting task>
```

### Issue Description

In the example, the user is trying to assign a value to a property, but `NDFrame.__setattr__` first calls `object.__getattribute__` on the same name. It wastes runtime by unnecessarily calling properties or raising `AttributeError`. Here is the original code:
```
    @final
    def __setattr__(self, name: str, value) -> None:
        """
        After regular attribute access, try setting the name
        This allows simpler access to columns for interactive use.
        """
        # first try regular attribute access via __getattribute__, so that
        # e.g. ``obj.x`` and ``obj.x = 4`` will always reference/modify
        # the same attribute.

        try:
            object.__getattribute__(self, name)
            return object.__setattr__(self, name, value)
        except AttributeError:
            pass

        # if this fails, go on to more involved attribute setting
        # (note that this matches __getattr__, above).
        if name in self._internal_names_set:
            object.__setattr__(self, name, value)
        elif name in self._metadata:
            object.__setattr__(self, name, value)
        else:
            try:
                existing = getattr(self, name)
                if isinstance(existing, Index):
                    object.__setattr__(self, name, value)
                elif name in self._info_axis:
                    self[name] = value
                else:
                    object.__setattr__(self, name, value)
            except (AttributeError, TypeError):
                if isinstance(self, ABCDataFrame) and (is_list_like(value)):
                    warnings.warn(
                        "Pandas doesn't allow columns to be "
                        "created via a new attribute name - see "
                        "https://pandas.pydata.org/pandas-docs/"
                        "stable/indexing.html#attribute-access",
                        stacklevel=find_stack_level(),
                    )
                object.__setattr__(self, name, value)


```

### Expected Behavior

The attribute should not be accessed, avoiding otherwise wasted time or unexpected behavior. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.11.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 6.1.69-1-MANJARO
Version          : #1 SMP PREEMPT_DYNAMIC Thu Dec 21 12:29:38 UTC 2023
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 2.0.3
numpy            : 1.24.4
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 23.2.1
Cython           : 0.29.36
pytest           : 7.4.0
hypothesis       : None
sphinx           : 7.1.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.14.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.2
numba            : 0.57.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.11.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : 0.9.0
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by dhodczak**
- [x] closes #56793
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6296,44 +6296,44 @@ def __getattr__(self, name: str):
...
+        if (
+            # only relevant with dataframes
+            isinstance(self, ABCDataFrame)
+            # ignore if attr already exists
+            and name not in self.__dict__
+            # ignore if it's a class attribute;
+            # might be a prop, axis, cached_property,
+            # or other sort of descriptor
+            and not hasattr(type(self), name)
+            # ignore if internal or metadata
```
Comment (rhshadrach) [inline pandas/core/generic.py:6324]: These comments are almost verbatim repetitions of the code; I think they can be removed.

**Review by rhshadrach — COMMENTED** [review]
As mentioned in the issue, the use case for subclassing pandas objects where a given attribute may be doing computation when accessed makes sense to avoid. I also like less try-excepts here. That said - this also appears to me it might make the code more fragile. An update to `__getattribute__` not mirrored here could break things.

But overall still positive here. cc @phofl

**Comment by dhodcz2** [comment]
The previous commit passed and despite only committing the removal of commits, the latest commit fails workflow:
```
=========================== short test summary info ============================
FAILED pandas/tests/generic/test_to_xarray.py::TestDataFrameToXArray::test_to_xarray_index_types[string] - FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.
FAILED pandas/tests/generic/test_to_xarray.py::TestDataFrameToXArray::test_to_xarray_index_types[datetime] - FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.
FAILED pandas/tests/generic/test_to_xarray.py::TestDataFrameToXArray::test_to_xarray_index_types[datetime-tz] - FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.
```
It seems that the workflow environment has updated, but the old test remains the same, using `Dataset.dims` instead of `Dataset.sizes`. This causes a failure in my 3-week old commit. How should I proceed?

**Comment by dhodcz2** [comment]
I'm still interested in finalizing this pull request. If there are any additional changes needed or further reviews required, please let me know and I'll address them promptly.

**Comment by mroeschke** [comment]
Thanks for the PR but it has appears to gone stale. Additionally, this probably needs more discussion on the issue regarding the fix so closing until that discussion has resolved

**Pull request closed without merging**